### DISCUSSION TITLE
Is there an equivalent to postgres pg_dump --schema-only?

### DISCUSSION THREAD

**Discussion by rgl** [ClickHouse/ClickHouse#60382] [Q&A]
Is there an equivalent to postgres `pg_dump --schema-only`?

As described at https://www.postgresql.org/docs/current/app-pgdump.html, `--schema-only`, dumps all the database schema as SQL statements (e.g. `create table` statements).

I've tried using `clickhouse client --query 'select name, create_table_query from system.tables where database='uptrace' format Vertical;'` but that is a too verbose query and far from being something that could be put in a `.sql` file and execute it to recreate the schema in a fresh database. And it seems to lack other database objects, like columns.

**Comment by mlex** [comment]
If you use a different output format and just export the  create_table_query column then the result is valid sql (and consumably by clickhouse). For example:

```
clickhouse client --host ... --query "select create_table_query from system.tables where database NOT IN ('system', 'information_schema', 'INFORMATION_SCHEMA') format TabSeparatedRaw"

CREATE TABLE default.test (`x` Int32, `y` String) ...
```

**Comment by rgl** [comment]
Thank you for trying, but I'm afraid that is too simple. I'm looking for a built-in way, just like `pg_dump`. Without something like that, we, the users, we'll play the whack-a-mole game about which database objects go in a proper schema .sql file. If this feature does not yet exist, I'll happily open an github issue about it.

**Reply by mlex** [reply]
> If this feature does not yet exist, I'll happily open an github issue about it.

To my knowledge, there is no such feature as a single command.

**Reply by tsolodov** [reply]
you can find table definition in data directory:

```
find . -name *.sql -type f
./metadata/default.sql
./metadata/information_schema.sql
./metadata/system.sql
./store/4d6/4d654e2d-3a54-4adc-bc32-d91577ff88b6/test_table.sql
```

one problem, the sql files do `ATTACH TABLE`, not `CREATE TABLE` + table name is part of filename, maybe it helps...

**Reply by tsolodov** [reply]
another way(but it dumps netadata + data):
```
:) BACKUP table wikistat TO File('/tmp/dddd.sql')

BACKUP TABLE wikistat TO File('/tmp/dddd.sql')

Query id: 1e33fb8e-1cb6-4a8f-b4e5-eecab083d40b

┌─id───────────────────────────────────┬─status─────────┐
│ 5031f257-8a85-4fa7-a30b-d95283822013 │ BACKUP_CREATED │
└──────────────────────────────────────┴────────────────┘

1 row in set. Elapsed: 0.003 sec.
```

```
cat /tmp/dddd.sql/metadata/default/wikistat.sql
CREATE TABLE default.wikistat (`time` DateTime CODEC(Delta(4), ZSTD(3)), `project` LowCardinality(String), `subproject` LowCardinality(String), `path` String CODEC(ZSTD(3)), `hits` UInt64 CODEC(ZSTD(3))) ENGINE = MergeTree ORDER BY (path, time) SETTINGS index_granularity = 8192
```

**Reply by rgl** [reply]
Indeed, using the [`backup query/command`](https://clickhouse.com/docs/en/operations/backup) is a workaround.

Using https://github.com/Altinity/clickhouse-backup with the `--schema` parameter is also a workaround.

But neither is as convenient to use as `pg_dump --schema-only`.

**Reply by qoega** [reply]
Why you can't use `SELECT create_table_query FROM system.tables` ?

**Reply by rgl** [reply]
@qoega from what I understood, that only includes tables, but there are other database objects that are part of a "schema", like, indexes, views, etc.

**Reply by qoega** [reply]
There is the similar system.views for them etc.

**Reply by rgl** [reply]
@qoega and there lies the problem with that ad-hoc approach to implement something like `pg_dump --schema` for ch. hence https://github.com/ClickHouse/ClickHouse/issues/60681.

**Reply by qoega** [reply]
I do not see any problem writing query with 3 system tables and UNION ALL. There is no need in tool for that purpose, but BACKUP command may have metadata only option for convenience

**Reply by rgl** [reply]
@qoega see it from another viewpoint, please follow this use-case. assume you are a ch casual user, which does not have inner knowledge about ch, nor you follow ch on a daily basis. you want to peek the schema of an unfamiliar software that uses ch. which you also do not have inner knowledge. even so, you want to bootstrap your knowledge about the two. so you start with some expectations, since you are already familiar with other databases systems, so you take a leap of faith, and you expect to find a tool to dump a database schema and data, preferable without jumping too many hoops, something like `pg_dump --schema`. then you start to look around, and come up with the `backup` sql statement/command, which is not ideal for the above use-case, which makes you jump too many hoops. hence this plea for a more straightforward tool.

**Comment by rgl** [comment]
I've created the feature request at https://github.com/ClickHouse/ClickHouse/issues/60681.

**Comment by abraithwaite** [comment]
Would love to have this be a native part of ClickHouse, but in the meantime I've written this small utility program if it's helpful to anyone:

https://github.com/runreveal/chdump
